### ISSUE TITLE
Events review

### ISSUE THREAD

**Issue by simonziegler** [Material-Blazor/Material.Blazor#365]
Input from @stefanloerwald

**Comment by stefanloerwald** [comment]
Thanks for creating this. I'll be in touch soon-ish. It's the festive season after all!

**Comment by stefanloerwald** [comment]
Let me briefly describe the situation and how I intend to solve it.

### Premise

- There is a long long list of events that could potentially be used by _someone_ and it is tedious to implement them all explicitly for all components. We need a easy-to-maintain and best-performance approach for a top-notch component library.

### Current implementation

- The current approach to handle them through the `CaptureUnmatchedValues` parameter in the component foundation (https://github.com/Material-Blazor/Material.Blazor/blob/9fc4b5e5ff6817147b812de2ecded6b29d66ed97/Material.Blazor/Foundation/ComponentFoundation.cs#L32)
- According to https://github.com/dotnet/AspNetCore.Docs/blob/master/aspnetcore/blazor/webassembly-performance-best-practices.md#avoid-attribute-splatting-with-captureunmatchedvalues, there is a performance overhead to using `CaptureUnmatchedValues`. See also https://github.com/dotnet/aspnetcore/discussions/27059
- Another issue of this approach is that `@onclick="Handler" @onclick:stopPropagation` does not work, see https://github.com/dotnet/aspnetcore/issues/20560.

### Thoughts & Proposed solution

- This is not yet fully thought through, so please be gentle when ripping this apart ;-)
- Not every component needs every event. A very long list of parameters is not a nice API, this needs to be tidy. So limiting to a sensible event set per component would be nice, I think.
- `stopPropagation` and `preventDefault` should be available wherever appropriate (I don't think those two apply to each and every event, but this would need to be verified)
- It may be the case that different events should be mapped to different dom elements in some components, who knows?
- Getting closer to a solution now....
- Source Generators have become mature enough to use them. It is trivial to add `[Parameter]` properties to components based on e.g. an attribute detected in a file. Ideally, this generated parameter also comes with some documentation, so that should be included in the generated code too.
- This eliminates tedious repetitive code for handling events for declaring the properties, but not for adding it to the DOM. A small set of helper-RenderFragments might be useful to map them. (I think I want this solution)
- Alternatively, The source generator can come with an analyzer that warns the developer if an event property is not used within the `BuildRenderTree` method.
- A very early implementation of this (generator+analyzer, not the helper RF) can be found here: https://github.com/excubo-ag/Generators.Blazor/blob/main/Blazor/EventParameterGenerator.cs
- In this prototype, you would add `[GenerateEvents(Mouse.All), GenerateEvents(Touch.All)]` to your component and it adds the relevant event properties, plus the `stopPropagation` and `preventDefault` boolean properties. (All event groups available here: https://github.com/excubo-ag/Generators.Blazor/blob/main/Blazor/Attributes.cs)
- The remaining development work would then be actually mapping the event parameters to the right DOM element(s). Adding more support for events is trivial, as that's just another attribute + using them.
- My vision includes the aforementioned helper fragment, where usage looks something like
```
@code {
   // this is something to be developed by implementing a source generator for best performance
   // classes, styles etc need to be kept in mind here as well, so careful thinking before rushing to an implementation!
   [GenerateEventHelper(Mouse.All), GenerateEventHelper(Touch.All), GenerateEventHelper("div")]
   partial RenderFragment<RenderFragment> DivWithEvents();
}

@DivWithEvents(@<text>
   <!--Child content of this div goes here as usual-->
</text>)
```

That's it for now ;-)

**Comment by simonziegler** [comment]
This is well worth thinking through. My sense is that we should experiment, do so slowly and if something good drops out, it can go in v3 because v2 preview is about to be released.

One thing. We also capture unmatched values to allow consumers to apply other tags including setting class and style tags. Let's bear this in mind (and consider whether it's actually needed) as we go.

Another thing. This is about performance. How do we assess Mareeial.Blazor performance anyway? And what do we consider to be problematic?

🎄🎁❄️☃️ 🎄🎁❄️☃️ 🎄🎁❄️☃️

**Comment by stefanloerwald** [comment]
In my opinion, performance is not a huge concern, as any major problem would have been noticable already. So my goal is more to allow `stopPropagation` and `preventDefault`.

**Comment by simonziegler** [comment]
### Update

Stefan is going to work gently on a potential re-engineering of how attributes and events are applied to MB components. This will probably result in removing the attribute splatting mechanism and using source code generation in some form.

**Comment by stefanloerwald** [comment]
I thought a bit more about this. Given that .NET6 will bring a new feature which will [enable users to write custom events](https://devblogs.microsoft.com/aspnet/asp-net-core-updates-in-net-6-preview-2/#support-for-custom-event-arguments-in-blazor), it doesn't seem wise to handle events any other way than with attribute splatting.

So I'm leaning towards leaving the event handling code as is (after #421). Which leaves the issue that `stopPropagation`/`preventDefault` isn't supported directly on the components. I have yet to see a case where `preventDefault` would be necessary for an MB component. and I would imagine `stopPropagation` could always be stopped by wrapping the component, e.g.

```html
<div @onclick:stopPropagation>
   <MBButton @onclick="Click" Label="Click me" />
</div>
```

An unwanted side effect of handling events as parameters would also be the silent(!!) breaking change for users `@onclick` -> `onclick`.

So I vote for closing this topic for now.
